universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java

```diff
@@ -52,11 +54,17 @@ public Result index(long programId) {
   }
 
   @Secure(authorizers = Authorizers.Labels.UAT_ADMIN)
-  public Result create(long programId) {
+  public Result create(Request request, long programId) {
     try {
-      ProgramDefinition program = programService.addBlockToProgram(programId);
-      long blockId = program.getLastBlockDefinition().id();
-      return redirect(routes.AdminProgramBlocksController.edit(programId, blockId).url());
+      ErrorAnd<ProgramDefinition, CiviFormError> result =
+          programService.addBlockToProgram(programId);
+      ProgramDefinition program = result.getResult();
+      BlockDefinition block = program.getLastBlockDefinition();
+      if (result.isError()) {
+        String errorMessage = joinErrors(result.getErrors());
+        return renderEditViewWithMessage(request, program, block, errorMessage);
+      }
```

Comment:
suggestion(blocking): I think this error check needs to be higher - we try to get the result on line 61, which will throw a RuntimeException if there is no result

Reply:
The result actually includes an unmodified program definition in the error case. We need the program definition anyway to finish the rendering regardless whether the add succeeds or not so I figured I can return it with the errors. I can update the docstring to make that clearer. Or if you feel this is making it more confusing than the benefit of convenience, I can change it back to have a second call to `getProgramDefinition` in the error case.

Reply:
Updated javadoc.

Reply:
Thank you, that makes much more sense! Thank you for updating the documentation, too